File: backend/sidegames/blackjack.py
"""Blackjack against the dealer: the arithmetic, and nothing else.

Twenty-one is a game with very few decisions in it and a great many edge cases,
and almost every edge case is an ace. An ace is worth eleven until eleven would
bust the hand and then it is worth one — which sounds like a single sentence and
is the reason this module exists on its own. An ace counted as eleven when it
had to be one is somebody's coins, so the rule lives in one function that a test
can hammer rather than being spelled out again at each of the five places that
need it.

Everything here is arithmetic over lists of card strings in the engine's format:
a rank character and a suit letter, "Th", "As", "9s". No database, no deck kept
between calls, no state that survives a return. The shuffle takes its generator
as an argument exactly the way spingo.draw_multiplier does, so a test can stack
the deck and know what the next card is; without one it uses the module-level
`random`, seeded from the OS.

The house rules below are decisions rather than laws, and each of them costs or
saves the house something:

* One 52-card deck, reshuffled every round. Nothing here can be counted, which
  is the point — counting a single deck that is thrown away after one hand is
  not a strategy.
* The dealer stands on soft 17. The kinder of the two standard rules, and the
  one that makes the game explicable: the dealer's whole policy is then "draw
  under seventeen", with no second clause about what kind of seventeen it is.
* A blackjack pays 3:2, rounded down. Rounding down rather than up because the
  alternative is a stake of 5 paying 8 back on a 7.5, and a house that rounds in
  the player's favour on every odd stake is a faucet nobody meant to open.
* Doubling takes exactly one card, and the hand is then over whatever it came
  to. That is what makes it a double and not a licence to keep drawing.
* Splitting is allowed on equal RANK only. K+Q is twenty either way, but it is
  not a pair, and a house that lets you split it is giving away the difference.
* No re-splitting — two hands is the maximum — and split aces get one card each
  and are done. Both of those are limits on the same thing: an ace is the best
  card in the deck and the split is the cheapest way to get more of them.
* No insurance and no surrender. Insurance is a side bet on the dealer's hole
  card that is bad for the player at these odds, and explaining it costs more
  than it is worth.

Two more that the contract does not name either way, and that were decided here
because something had to be:

* A hand that reaches twenty-one stands rather than staying live. Every card in
  the deck busts a twenty-one, so a Hit button offered on one is a trap and not
  a choice. See status_after_card.
* The dealer is checked for blackjack on the deal — the peek — and a round it
  finds one in is over before the player has acted. Without it a player doubles
  or splits into a hand that was lost before they touched it and pays a second
  stake to find out. The check itself is in blackjackbank.deal, where the cards
  are; the rule is here because it is a rule.
"""

import random

# The card vocabulary, in the format game/engine/card.py prints. Ten is "T" so
# that every card is exactly two characters and a hand is a list of equal-width
# strings rather than a parsing problem.
RANKS = "23456789TJQKA"
SUITS = "hdcs"

# What the dealer's hole card is called on the wire while the player is still
# deciding. It never reaches any function in this module: the hiding happens
# when the round is drawn for the client, not in the round itself, so nothing
# here ever has to wonder what a "??" is worth.
HIDDEN = "??"

TWENTY_ONE = 21

# The dealer draws below this and stands on it, soft or hard. See the note at
# the top: a house that hits soft 17 would read `total < 17 or (total == 17 and
# soft)`, and the extra clause is the whole difference between the two games.
DEALER_STANDS_ON = 17

# A blackjack pays three for two, as a pair rather than as 1.5 — the payout is
# integer arithmetic and a float would have introduced a rounding rule nobody
# chose.
BLACKJACK_PAYS = (3, 2)

# Two hands, never three. See the note at the top about re-splitting.
MAX_HANDS = 2

# What a hand is doing. "blackjack" is a status and not only an outcome because
# a natural stops the hand where it stands: there is nothing to decide on it and
# it must not be offered a hit.
PLAYING = "playing"
STOOD = "stood"
BUST = "bust"
BLACKJACK = "blackjack"

# How a hand ended up against the dealer. "blackjack" is separate from "win"
# because it is paid differently and because the client says so out loud.
WIN = "win"
LOSE = "lose"
PUSH = "push"

# What a round is doing. The same two words the API contract uses.
FINISHED = "finished"

NO_ACTIONS = {"hit": False, "stand": False, "double": False, "split": False}


def fresh_deck(rng=None) -> list:
    """A shuffled 52, as card strings, dealt from the front.

    A new one every round rather than a shoe that is dealt down: a round is the
    unit of everything here — one row, one stake, one settlement — and a deck
    that outlived it would be state the player could reason about.

    The generator is an argument so a test can stack it. Without one this uses
    the module-level `random`, which is not cryptography but must not be
    predictable from the last hand either.
    """
    rng = rng or random
    deck = [rank + suit for rank in RANKS for suit in SUITS]
    rng.shuffle(deck)
    return deck


def draw(deck: list) -> str:
    """Take the top card, off the front, mutating the deck it came from.

    The front rather than the back so that a stacked deck reads in dealing
    order in a test — `["As", "Kd", ...]` is the ace first, which is what
    anybody writing the test meant.
    """
    return deck.pop(0)


def card_rank(card: str) -> str:
    """The rank character of a card: "T" of "Th"."""
    return (card or "")[:1]


def card_value(card: str) -> int:
    """What one card is worth, counting every ace as eleven.

    Eleven rather than one because hand_value below is the only place that
    knows how to change its mind, and it needs a starting position to change it
    from. A picture is ten; there is no card worth anything else.
    """
    rank = card_rank(card)
    if rank == "A":
        return 11
    if rank in "TJQK":
        return 10
    return int(rank)


def hand_value(cards) -> tuple:
    """(total, soft) — what the hand is worth and whether an ace is doing it.

    Every ace starts as eleven and is demoted to one, one at a time, only while
    the hand is over twenty-one. That ordering is the whole rule: a hand is
    always worth the most it can be worth without busting, and demoting an ace
    that did not need demoting would quietly cost the player ten.

    "Soft" means an ace is still being counted as eleven, which is what makes
    the hand safe to hit — a soft 17 cannot bust on the next card. A+6 is a soft
    17; A+6+K is a hard 17, because by then the ace has had to become a one.
    """
    total = 0
    # Aces still being counted as eleven. This is the number that gets spent.
    high_aces = 0
    for card in cards:
        value = card_value(card)
        if card_rank(card) == "A":
            high_aces += 1
        total += value

    while total > TWENTY_ONE and high_aces:
        total -= 10
        high_aces -= 1

    return total, high_aces > 0


def total_of(cards) -> int:
    """The hand's value, for the callers that do not care how it got there."""
    return hand_value(cards)[0]


def is_bust(cards) -> bool:
    return total_of(cards) > TWENTY_ONE


def is_blackjack(cards, from_split: bool = False) -> bool:
    """An ace and a ten-card, as the first two cards, in a hand nobody split.

    Two conditions, and the second one is the one that gets forgotten: twenty-
    one made of three cards is twenty-one and pays even money, and twenty-one
    made from a split ace is twenty-one too. Paying 3:2 on either is paying for
    a hand the player was not dealt.
    """
    return not from_split and len(cards) == 2 and total_of(cards) == TWENTY_ONE


def is_natural(hand: dict) -> bool:
    """is_blackjack, for a hand dictionary rather than a list of cards."""
    return is_blackjack(hand.get("cards") or [], hand.get("from_split", False))


def dealer_should_hit(cards) -> bool:
    """Whether the dealer draws again. Stands on soft 17 — see DEALER_STANDS_ON.

    The dealer has no choices, which is why the house edge is where it is: it
    must draw on a hard 16 against a six even though a player never would, and
    it must stand on a soft 17 even when standing cannot win.
    """
    return total_of(cards) < DEALER_STANDS_ON


def play_dealer(cards: list, deck: list) -> list:
    """Draw for the dealer until the policy stops, mutating both lists."""
    while dealer_should_hit(cards):
        cards.append(draw(deck))
    return cards


def dealer_must_play(hands) -> bool:
    """Whether the dealer has anything left to draw for.

    Not when every hand went bust: those are already lost, the coins are already
    the house's, and cards dealt afterwards would only look like the house was
    chasing a hand it had won. Not when the only hand is a natural either — by
    the time this is asked the dealer has been checked for blackjack and has
    none, so a natural is already paid and what the dealer would have drawn to
    cannot change it.
    """
    return any(not is_bust(hand["cards"]) and not is_natural(hand) for hand in hands)


def new_hand(cards, stake: int, from_split: bool = False) -> dict:
    """One player hand, in the shape it is stored and served in.

    Built here rather than inline at the three places that make one, so the
    shape is defined once — a hand that reached the client missing `returned`
    would be a hand the client draws a zero against forever.
    """
    return {
        "cards": list(cards),
        # What this hand alone is playing for. A doubled hand carries the
        # doubled figure, because that is what it wins or loses.
        "stake": int(stake),
        "doubled": False,
        "from_split": bool(from_split),
        "status": PLAYING,
        "outcome": None,
        "returned": 0,
    }


def status_after_card(cards, one_card_only: bool = False) -> str:
    """What a hand becomes the moment a card lands on it.

    Twenty-one ends the hand rather than leaving it live. There is nothing to
    decide on twenty-one — every card in the deck busts it — and a Hit button
    offered on a hand that cannot survive being hit is a trap rather than a
    choice.

    `one_card_only` is the double: the hand is over at whatever it came to, even
    a twelve, because that is what doubling buys.
    """
    total = total_of(cards)
    if total > TWENTY_ONE:
        return BUST
    if one_card_only or total == TWENTY_ONE:
        return STOOD
    return PLAYING


def actions_for(hands, index: int, active) -> dict:
    """What this hand may legally do — the `can` object the client is served.

    The server's word, and the only word: an action arriving on an endpoint is
    checked against this again rather than trusted, because a client that offers
    a fifth card on a doubled hand is either broken or somebody's script.

    Everything is false for a hand that is not the one being played, which
    includes every hand once the round is over.
    """
    if index != active or index >= len(hands):
        return dict(NO_ACTIONS)

    hand = hands[index]
    if hand.get("status") != PLAYING:
        return dict(NO_ACTIONS)

    cards = hand["cards"]
    # Doubling and splitting are both "on the first two cards" — once a hand has
    # been hit it is an ordinary hand that can only be hit again or stood.
    untouched = len(cards) == 2
    return {
        "hit": True,
        "stand": True,
        "double": untouched,
        # Equal rank, and no re-splitting: with two hands already in play there
        # is no room for a third, whichever of them is holding the pair.
        "split": (
            untouched
            and len(hands) < MAX_HANDS
            and card_rank(cards[0]) == card_rank(cards[1])
        ),
    }


def returns_for(outcome: str, stake: int) -> int:
    """Coins back to a hand, the stake included.

    Returned rather than won, because the stake left the wallet when the cards
    were dealt: a push pays the stake back and is worth nothing, and saying so
    in one number is what lets the round report a net without the client doing
    arithmetic of its own.
    """
    stake = max(0, int(stake))
    if outcome == BLACKJACK:
        # Three for two on top of the stake. Floor division: see BLACKJACK_PAYS.
        numerator, denominator = BLACKJACK_PAYS
        return stake + stake * numerator // denominator
    if outcome == WIN:
        return stake * 2
    if outcome == PUSH:
        return stake
    return 0


def settle(hands, dealer_cards) -> list:
    """How each hand ended against the dealer, and what it gets back.

    Reads the cards rather than the statuses. A status is bookkeeping written by
    whichever endpoint last touched the round; the cards are what was actually
    dealt, and if the two ever disagree it is the cards that were paid for.

    The order of the clauses is the rule set, and it is worth reading as one:
    a bust hand has already lost and the dealer's hand is none of its business,
    which is the entire house edge. Then the dealer's blackjack, which beats
    every twenty-one that was not itself a blackjack and pushes against one that
    was. Then everybody else, high hand wins, equal is a push.
    """
    dealer_total = total_of(dealer_cards)
    dealer_blackjack = is_blackjack(dealer_cards)
    dealer_bust = dealer_total > TWENTY_ONE

    results = []
    for hand in hands:
        cards = hand["cards"]
        if is_bust(cards):
            # Lost the moment it happened, whatever the dealer goes on to do.
            outcome = LOSE
        elif is_natural(hand):
            outcome = PUSH if dealer_blackjack else BLACKJACK
        elif dealer_blackjack:
            outcome = LOSE
        elif dealer_bust or total_of(cards) > dealer_total:
            outcome = WIN
        elif total_of(cards) == dealer_total:
            outcome = PUSH
        else:
            outcome = LOSE
        results.append({"outcome": outcome, "returned": returns_for(outcome, hand["stake"])})
    return results

File: frontend/src/components/game/blackjack.js
/**
 * What the blackjack screen says, and which of its buttons do anything.
 *
 * All of this module is formatting and offering. None of it decides what is
 * true about a hand. The server deals, counts and settles, and the round it
 * sends back is the only account of the game anybody is actually paid out of —
 * so the totals here are read off the payload rather than added up from the
 * cards, and the buttons come from the server's own `can` rather than from the
 * house rules restated in JavaScript. A client that quietly disagreed about a
 * soft 17 would not change what the hand was worth; it would just tell somebody
 * they had won while their balance said otherwise, which is the one bug in a
 * game like this that costs real coins and all of somebody's trust.
 *
 * Pure, and tested, for the same reason the tier rows are: every line of it is
 * about money — what a bet costs, whether the wallet covers it, what a finished
 * hand paid — and those are exactly the sentences that sit wrong inside JSX for
 * a month and are obvious in a test.
 */

// The house limits, for a screen drawn before the wallet's game list has
// arrived. They are the same numbers the server enforces (sidegames/games.py);
// the point of the fallback is only that the betting screen can be drawn at
// all, not that the client gets an opinion about the limits.
const FALLBACK_MIN = 5;
const FALLBACK_MAX = 500;

/** Coins the way the rest of the app prints them: figures, with separators. */
const coins = (amount) => Number(amount || 0).toLocaleString();

/** The card the dealer has not turned over yet, as the engine writes it. */
export const HIDDEN_CARD = "??";

/**
 * The chips a bet is built from, smallest first — the order they sit in as a
 * row of buttons under the felt.
 *
 * Three denominations, not the table's seven: stakes here run 5 to 500 and
 * every one of them is a small handful of these. More chips would be more to
 * read and no bet you could not already make.
 */
export const CHIPS = [5, 25, 100];

/**
 * An amount as the pile of chips that makes it, largest first.
 *
 * Greedy, which for 5/25/100 is also exact — every multiple of five up to the
 * table maximum comes out in the fewest chips there are. Largest first because
 * that is the order a pile is built in front of a player: the hundred goes down
 * and the change goes on top, and a pile stacked the other way up reads as
 * somebody else's bet.
 *
 * Never returns a denomination the table does not have, so whatever is left of
 * an odd amount is simply not drawn. The figure is printed beside the pile
 * anyway, and inventing a chip to cover the remainder would be the drawing
 * telling a small lie about what is on the felt.
 */
export function chipsFor(amount) {
  const wanted = Number(amount);
  if (!Number.isFinite(wanted) || wanted <= 0) return [];

  const pile = [];
  let left = Math.floor(wanted);
  for (const chip of [...CHIPS].reverse()) {
    while (left >= chip) {
      pile.push(chip);
      left -= chip;
    }
  }
  return pile;
}

/**
 * The stakes this table takes, as {min, max}.
 *
 * Reads both spellings on purpose. The wallet sends `min_stake`/`max_stake`
 * with each game, and a caller with limits of its own says `min`/`max`; a
 * helper that understood only one of them would silently hand back the
 * fallbacks for a game row it was given in good faith, which is a bet quietly
 * measured against the wrong table.
 */
export function stakeLimits(limits) {
  return {
    min: Number(limits?.min ?? limits?.min_stake ?? FALLBACK_MIN),
    max: Number(limits?.max ?? limits?.max_stake ?? FALLBACK_MAX),
  };
}

/**
 * Whether an opening bet is one this table and this wallet will take.
 *
 * Only the opening bet. A double or a split takes a *second* stake off the
 * wallet (see the contract: a 500 split costs 1,000 in all), so a bet that
 * passes here is still no promise that the hand can be doubled — that question
 * is canCoverSecondStake, and the two are separate because they are asked at
 * different moments about different money.
 */
export function canAfford(amount, balance, limits = {}) {
  const { min, max } = stakeLimits(limits);
  const bet = Number(amount);
  if (!Number.isFinite(bet) || bet < min || bet > max) return false;
  return bet <= Math.max(0, Number(balance) || 0);
}

/**
 * Whether the wallet still holds a second stake for this hand.
 *
 * Doubling and splitting are not free moves, and the player most likely to
 * reach for them is the one who just pushed everything they had into the
 * opening bet. The server refuses it anyway, but a button that lights up and
 * then errors costs somebody the hand they were in the middle of playing.
 *
 * Measured against the hand rather than the round because a doubled hand
 * reports the doubled stake, and it is the hand in front of the player that the
 * next stake is being paid for.
 */
export function canCoverSecondStake(hand, balance) {
  const stake = Number(hand?.stake) || 0;
  if (stake <= 0) return false;
  return Math.max(0, Number(balance) || 0) >= stake;
}

/**
 * What the betting screen says and offers, before there are any cards.
 *
 * `game` is the row out of the wallet's game list, which carries the limits;
 * it is often not there yet on the first paint, so the house numbers stand in.
 *
 * When Deal is off it says why, and the order the reasons are checked in is the
 * point of the function. The table maximum is a fact about the bet and is true
 * for everybody, so it is answered first: telling a player they are short of
 * coins for a bet this table would refuse from anyone sends them to the shop
 * for nothing. A bet of zero is not a mistake and is not told off for being
 * one: nobody has done anything yet, so the reason is the instruction — and
 * `reason` is what the dead Deal button says on its face, so it has to read as
 * a sentence somebody would put on a button rather than as a complaint.
 */
export function bettingState({ bet = 0, balance = null, game = null } = {}) {
  const { min, max } = stakeLimits(game);
  const purse = Math.max(0, Number(balance) || 0);
  const amount = Math.max(0, Math.floor(Number(bet) || 0));

  let reason = null;
  if (amount === 0) reason = "Place a bet";
  else if (amount > max) reason = `Table maximum is ${coins(max)}`;
  else if (amount > purse) reason = "Not enough coins";
  else if (amount > 0 && amount < min) reason = `Minimum bet is ${coins(min)}`;

  return {
    bet: amount,
    label: coins(amount),
    chips: chipsFor(amount),
    min,
    max,
    canDeal: reason === null,
    reason,
    canClear: amount > 0,
    // A chip that cannot be added is drawn dead rather than removed: the row of
    // chips is the same row all evening, and one that reshuffles as the bet
    // grows is a row you have to re-read every time you touch it.
    chipButtons: CHIPS.map((value) => ({
      value,
      label: coins(value),
      enabled: amount + value <= max && amount + value <= purse,
    })),
  };
}

/**
 * What to print for a hand's total.
 *
 * A soft hand is genuinely two hands until somebody draws, so it says both
 * readings — "7 / 17" is the whole of why you might hit a seventeen. The low
 * one is the high one minus the ace's ten; that is arithmetic on the server's
 * own number, not a recount of the cards.
 *
 * `total` and `soft` are taken as given. Recomputing them from the cards would
 * be inviting the client to disagree with the hand it is being paid on.
 */
export function handTotal(hand) {
  if (!hand) return "";
  const total = Number(hand.total) || 0;
  if (hand.soft && total >= 11) return `${total - 10} / ${total}`;
  return String(total);
}

/**
 * The short word over a hand, or null while it is simply being played.
 *
 * 21 outranks "Stood" because it is the better news and the more useful: a
 * player glancing along two split hands wants to know which one is the good
 * one. It is read off the total rather than the status so that a 21 says so
 * from the moment it happens, whether or not the server has stood it yet.
 */
export function handLabel(hand) {
  if (!hand) return null;
  if (hand.status === "blackjack") return "Blackjack";
  if (hand.status === "bust") return "Bust";
  if (Number(hand.total) === 21) return "21";
  if (hand.status === "stood") return "Stood";
  return null;
}

/**
 * What a settled hand says: how it went, and what it paid.
 *
 * The figure is the hand's own movement — what came back, less what went in —
 * so a blackjack on 25 reads "+37" rather than the 62 that landed in the
 * wallet. What somebody wants to know is what they made, and `returned` on its
 * own overstates it by the stake every time.
 *
 * A push and a loss carry no number. A push moved nothing, and the stake a loss
 * cost is the pile still sitting on the felt in front of them; the round's own
 * line underneath has the net for anybody who wants it in figures.
 */
export function outcomeLine(hand) {
  const outcome = hand?.outcome;
  if (!outcome) return null;

  const won = Number(hand.returned || 0) - Number(hand.stake || 0);
  if (outcome === "blackjack") return `Blackjack · +${coins(won)}`;
  if (outcome === "win") return `Won · +${coins(won)}`;
  if (outcome === "push") return "Push";
  return "Lost";
}

// How each outcome is spoken about after the fact, and the order two of them
// are read out in — the good news first, which is the order anybody tells it.
const PAST = { win: "won", lose: "lost", push: "pushed" };
const TELLING_ORDER = ["win", "lose", "push"];

/**
 * One line for the whole round, for the moment the player is told how it went.
 *
 * Two split hands can come out differently, and this is the line that has to be
 * honest about it: "One won, one lost" over a net of nothing is the truth, and
 * anything that rounded that to "Lost" or to "Won" would be picking a half of
 * the round to report. `tone` is taken from the net rather than from the words,
 * because what the colour is answering is "did that cost me anything" — and for
 * a split that came out both ways the answer really is no.
 */
export function roundSummary(round) {
  if (!round || round.status !== "finished") return null;

  const hands = round.hands || [];
  const net = Number(round.net) || 0;
  let tone = "push";
  if (net > 0) tone = "win";
  else if (net < 0) tone = "lose";

  return {
    headline: headlineFor(hands),
    net,
    netLabel: net === 0 ? null : `${net > 0 ? "+" : ""}${coins(net)}`,
    tone,
  };
}

/**
 * The words for however many hands were played.
 *
 * Two is the most there can ever be — the house does not re-split — so the
 * mixed case is only ever a pair, and it is spoken as a pair rather than as a
 * list of outcomes.
 */
function headlineFor(hands) {
  if (!hands.length) return "Round over";

  if (hands.length === 1) {
    const [hand] = hands;
    if (hand.outcome === "blackjack") return "Blackjack";
    if (hand.outcome === "win") return "Won";
    if (hand.outcome === "push") return "Push";
    // A bust is a loss, but it is the loss the player did to themselves, and
    // saying so is the difference between a result and an account of it.
    if (hand.status === "bust") return "Bust";
    if (hand.outcome === "lose") return "Lost";
    return "Round over";
  }

  // A split cannot be a blackjack (house rule), but a server that ever called
  // one that should not read as a fourth kind of outcome down here.
  const outcomes = hands.map((hand) => (hand.outcome === "blackjack" ? "win" : hand.outcome));
  const kinds = TELLING_ORDER.filter((kind) => outcomes.includes(kind));
  if (!kinds.length) return "Round over";
  if (kinds.length === 1) return `Both hands ${PAST[kinds[0]]}`;
  return `One ${PAST[kinds[0]]}, one ${PAST[kinds[1]]}`;
}

/** The hand being played, or null once the round belongs to the dealer. */
export function activeHand(round) {
  if (!round || round.active == null) return null;
  return (round.hands || [])[round.active] || null;
}

// Fixed, and in this order everywhere: a button that changes position between
// one hand and the next is a button somebody hits by accident, and every one of
// these four spends coins or ends a hand.
const ACTIONS = [
  { key: "hit", label: "Hit" },
  { key: "stand", label: "Stand" },
  { key: "double", label: "Double" },
  { key: "split", label: "Split" },
];

// The two that take another stake out of the wallet.
const SECOND_STAKE = ["double", "split"];

/**
 * The four buttons for the active hand, always all four and always in order.
 *
 * Dead ones stay in place rather than disappearing, so Stand is in the same
 * spot on a hand that cannot be split as on one that can.
 *
 * What is enabled comes from the server's `can` and never from the rules read
 * back off the cards — `can` is the server's word on what is legal, and the
 * client's job is to offer no more than it allows. `balance` is the one thing
 * added on top: the wallet is the client's own knowledge, and a double it
 * cannot pay for is better greyed out here than refused after the click. Leave
 * it out and the wallet is not consulted at all.
 */
export function actionButtons(round, { balance = null } = {}) {
  const hand = activeHand(round);
  const live = Boolean(hand) && round?.status !== "finished";
  const can = hand?.can || {};

  return ACTIONS.map(({ key, label }) => {
    let enabled = live && Boolean(can[key]);
    if (enabled && balance != null && SECOND_STAKE.includes(key)) {
      enabled = canCoverSecondStake(hand, balance);
    }
    return { key, label, enabled };
  });
}

/**
 * What the dealer's side of the table says, as the one line it is drawn as.
 *
 * While the hand is live the total is only of what is face up, and the line has
 * to say so: a "10" that becomes a 20 with no warning reads as the table
 * cheating, where "10 · one card down" is simply the game. The face-up ace
 * keeps both of its readings for the same reason a player's does — it is the
 * card the whole decision turns on.
 *
 * `revealed` is for the beat between the round settling and the hole card
 * actually turning over on screen. The settled total is withheld until then,
 * because the line sits beside the cards and one that announced 21 over a card
 * still face down would give away the only moment in this game worth watching.
 * The face-up total cannot be shown during that beat either: once the round is
 * finished the payload's total is the true one, and working the other out from
 * the cards would be the client counting a hand for itself.
 */
export function dealerLine(round, revealed = true) {
  const dealer = round?.dealer;
  if (!dealer) return "";

  if (round.status !== "finished") return `${handTotal(dealer)} \u00b7 one card down`;
  if (!revealed) return "one card down";

  const total = Number(dealer.total) || 0;
  if (dealer.blackjack) return `${total} \u00b7 Blackjack`;
  if (total > 21) return `${total} \u00b7 Bust`;
  return String(total);
}

/**
 * One hand of the last ten, as a mark on the strip.
 *
 * A letter rather than a word: ten of them have to fit across a phone, and the
 * strip is read as a shape — three reds in a row — before any of it is read as
 * text. The full story goes in the tooltip, where there is room for it.
 *
 * Blackjack gets its own mark rather than counting as a win. It paid 3:2 and it
 * is the best thing that happens in this game; a strip that flattened it into a
 * W would be hiding the only rows anybody wants to point at.
 */
export function historyMark(row) {
  const net = Number(row?.net) || 0;
  const figure = net > 0 ? `+${coins(net)}` : net < 0 ? `-${coins(Math.abs(net))}` : null;
  const MARKS = {
    blackjack: { label: "BJ", tone: "blackjack", word: "Blackjack" },
    win: { label: "W", tone: "win", word: "Won" },
    lose: { label: "L", tone: "lose", word: "Lost" },
    push: { label: "P", tone: "push", word: "Push" },
  };
  const mark = MARKS[row?.result] || MARKS.push;
  return {
    ...mark,
    // "Won · +50", or just "Push": a hand that moved nothing has no figure to
    // give, and "Push · 0" reads as a number somebody should care about.
    title: figure ? `${mark.word} \u00b7 ${figure}` : mark.word,
  };
}

/**
 * Whether the blackjack drawer should be on screen over a poker table.
 *
 * The one rule in this feature that can cost somebody real money, so it is a
 * function rather than a condition buried in a component: poker is the game and
 * blackjack is the waiting, and the moment the table needs you the card game is
 * not on screen. A player who misses a decision at a money table because
 * blackjack was covering it will not open it a second time.
 *
 * Deliberately not "hide it if they are in a hand" or "ask them first". It is
 * their turn or it is not, and nothing is lost by closing: the round lives on
 * the server and is still there when they fold again.
 */
export function drawerVisible({ open = false, isMyTurn = false } = {}) {
  return Boolean(open) && !isMyTurn;
}

/** Whether a card is the dealer's hole card, still face down. */
export function isHiddenCard(card) {
  return card === HIDDEN_CARD;
}
